walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,90 @@
 
 package com.google.sps;
 
+import static java.util.Collections.EMPTY_LIST;
+import static java.util.Collections.disjoint;
+import static java.util.Collections.sort;
+import static java.util.Arrays.asList;
+
+import java.util.Comparator;
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.List;
+
 
+// FindMeetingQuery finds open meeting timeslots throughout the day.
 public final class FindMeetingQuery {
+  /** 
+  *  @param events Set of events that attendees have, that need to be avoided.
+  *  @param request The duration of requested meeting and people attending. 
+  *  @return A list of open meeting timeslots.  
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+      List<Event> eventsList = new ArrayList<>(events);
+      sort(eventsList, new Comparator<Event>() {
+        public int compare (Event e1, Event e2) {
+            return TimeRange.ORDER_BY_START.compare(e1.getWhen(), e2.getWhen());
+          }
+        });
+      /** We want to remove any attendees not attending or events less-than/equal to 0. */
+      eventsList.removeIf(e -> (
+          e.getWhen().duration() <= 0 || disjoint(e.getAttendees(), request.getAttendees())));
+    
+      if( request.getAttendees().isEmpty()){
+          return asList(TimeRange.WHOLE_DAY);
+      }
+      if(request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+          return EMPTY_LIST;
+      }
+      if(eventsList.size()==0){
+          return asList(TimeRange.WHOLE_DAY);
+      }
+      if(eventsList.size()==1){
+          return asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, eventsList.get(0).getWhen().start(), false),
+          TimeRange.fromStartEnd(eventsList.get(0).getWhen().end(), TimeRange.END_OF_DAY, true)); 
+       }
+
+      return getOpenTimeSlots(eventsList, request);
+  }
+  /**
+  *  This method gets the open time slots by looping through a list of events
+  *  given that the list is sorted by start time. 
+  *  @param eventsList list of events in sorted order (by start time).
+  *  @param request he duration of requested meeting and people attending.
+  *  @return A list of open meeting timeslots. 
+  */
+  private Collection<TimeRange> getOpenTimeSlots(List<Event> eventsList, MeetingRequest request){
+      Collection<TimeRange> queryResult = new ArrayList<>();
+      for(int i=0;i<eventsList.size();i++){
+        /** 
+        *  First event in the list, check if there is enough time between event start and the start of the day. 
+        *  If there is enough time, add to the result. Since the events are sorted by start time we can check the
+        *  first index.
+        */
+        if(i==0){
```

Comment:
this doesn't need to be in the loop since it's a special case for the beginning of the list.

Reply:
Done.